### PR TITLE
Added v-grid api for fetching data items by row index

### PR DESCRIPTION
**Pull request by tomivirkki**
[Review on Reviewable](https://reviewable.io/reviews/vaadin/components/45)

### CODE REVIEW

File: vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java [hunk 1]
```diff
@@ -76,4 +80,38 @@ protected void resetDataAndSize(int newSize) {
...
+                            @Override
+                            public void onResponse(List<Object> rowData,
+                                    int totalSize) {
+                                JsUtils.jsni(callback, "call", callback,
+                                        JS.getUndefined(), rowData.get(0));
+                            }
+                        });
+            }
+        } else {
+            JS.exec(callback, JS.getError());
```
Comment (jounik) [inline vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java]: The new Error object should contain a message which describes the error.

File: vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java [hunk 2]
```diff
@@ -76,4 +80,38 @@ protected void resetDataAndSize(int newSize) {
...
+    public void getItem(Double rowIndex, JavaScriptObject callback,
+            boolean onlyCached) {
+        Integer index = JSValidate.Integer.val(rowIndex, -1, -1);
+        if (index >= 0 && index < size()) {
+            Object row = getRow(index);
+            if (row != null) {
+                JsUtils.jsni(callback, "call", callback, JS.getUndefined(),
+                        extractDataItem(row));
+            } else if (onlyCached) {
+                JS.exec(callback, JS.getError());
```
Comment (jounik) [inline vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java]: The new Error object should contain a message which describes the error.

**Comment by jounik** [comment]
Apart from the two ambiguous Error objects, LGTM.

**Comment by manolo** [comment]
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java, line 107 \[r1\]](https://reviewable.io:443/reviews/vaadin/components/45#-JtCieFcpa8D5CxxzVBA-r1-107)** ([raw file](https://github.com/vaadin/components/blob/638f0a14e8675b477490422a34600334062ef0dd/vaadin-components-gwt/src/main/java/com/vaadin/components/grid/data/GridDataSource.java#L107)):</sup>
Agree, can you add a parameter to getError(String message)  ?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/components/45)

**Comment by manolo** [comment]
Reviewed 9 of 11 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/components/45)

**Pull request merged**